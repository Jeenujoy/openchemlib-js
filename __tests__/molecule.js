'use strict';

const fs = require('fs');

const Molecule = require('../minimal').Molecule;

describe('Molecule', function () {
  it('fromSmiles', function () {
    testFromSmiles('C');
    testFromSmiles('COCOC');
    testFromSmiles('c1cc2cccc3c4cccc5cccc(c(c1)c23)c54');

    expect(function () {
      Molecule.fromSmiles('ABC');
    }).toThrow(/SmilesParser: unknown element label found/);

    function testFromSmiles(smiles) {
      const mol = Molecule.fromSmiles(smiles);
      expect(Molecule.fromSmiles(mol.toSmiles()).getIDCode()).toBe(
        mol.getIDCode()
      );
    }
  });

  it('medley', function () {
    const idcode =
      'enYXNH@MHDAELem`OCIILdhhdiheCDlieKDdefndZRVVjjfjjfjihJBbb@@@';
    let mol = Molecule.fromIDCode(idcode);

    const molfile = mol.toMolfile();
    mol = Molecule.fromMolfile(molfile);
    expect(mol.getIDCode()).toBe(idcode);

    const smiles = mol.toSmiles();
    mol = Molecule.fromSmiles(smiles);
    expect(mol.getIDCode()).toBe(idcode);
  });

  it('toSVG', function () {
    const mol = Molecule.fromSmiles('CCOCCO');
    let svg = mol.toSVG(300, 150, 'myId');
    expect(svg).toContain('width="300px" height="150px"');
    expect(svg).toContain('myId:Bond:0-1');
    expect(svg).toContain('myId:Atom:0');

    svg = mol.toSVG(300, 300, 'myId', {
      strokeWidth: 2,
      fontWeight: 'bold'
    });
    expect(svg).toContain('font-weight="bold"');
    expect(svg).toContain('stroke-width:2');
  });

  it('molfile V3', function () {
    const idcode =
      'enYXNH@MHDAELem`OCIILdhhdiheCDlieKDdefndZRVVjjfjjfjihJBbb@@@';
    let mol = Molecule.fromIDCode(idcode);

    let molfileV3 = mol.toMolfileV3();
    mol = Molecule.fromMolfile(molfileV3);
    expect(mol.getIDCode()).toBe(idcode);

    molfileV3 = mol.toMolfileV3();
    mol = Molecule.fromMolfile(molfileV3);
    expect(mol.getIDCode()).toBe(idcode);
  });

  it('fromMolfileWithAtomMap', () => {
    const molfile = fs.readFileSync(
      `${__dirname}/data/molfileWithHMap.mol`,
      'utf8'
    );
    const result = Molecule.fromMolfileWithAtomMap(molfile);
    expect(result.molecule).toBeInstanceOf(Molecule);
    expect(result.molecule.toSmiles()).toBe('CCCCC');
    expect(result.map).toHaveLength(result.molecule.getAllAtoms());
    expect(new Set(result.map).size).toBe(result.molecule.getAllAtoms());
  });

  it('addMissingChirality', () => {
    let mol = Molecule.fromSmiles('CC(Cl)CC');
    expect(mol.getChiralText()).toBe('unknown chirality');
    mol.addMissingChirality();
    expect(mol.getChiralText()).toBe('racemate');

    mol = Molecule.fromSmiles('CC(Cl)CC');
    mol.addMissingChirality(Molecule.cESRTypeOr);
    expect(mol.getChiralText()).toBe('this or other enantiomer');

    mol = Molecule.fromSmiles('CC(Cl)CC');
    mol.addMissingChirality(Molecule.cESRTypeAbs);
    expect(mol.getChiralText()).toBe('this enantiomer');
  });

  it.skip('getCanonizedIDCode', () => {
    let idcode = 'didH@@RYm^Fh@BHx@';
    let mol = Molecule.fromIDCode(idcode);
    expect(
      mol.getCanonizedIDCode(Molecule.CANONIZER_DISTINGUISH_RACEMIC_OR_GROUPS)
    ).toBe(idcode);
    mol.inventCoordinates();
    expect(
      mol.getCanonizedIDCode(Molecule.CANONIZER_DISTINGUISH_RACEMIC_OR_GROUPS)
    ).toBe(idcode);

    idcode = 'didH@@RYm^Fh@BHX@';
    mol = Molecule.fromIDCode(idcode);
    expect(
      mol.getCanonizedIDCode(Molecule.CANONIZER_DISTINGUISH_RACEMIC_OR_GROUPS)
    ).toBe(idcode);
    mol.inventCoordinates();
    expect(
      mol.getCanonizedIDCode(Molecule.CANONIZER_DISTINGUISH_RACEMIC_OR_GROUPS)
    ).toBe(idcode);
  });
});
