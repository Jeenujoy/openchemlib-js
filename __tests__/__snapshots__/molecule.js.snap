// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecule toSVG with autoCrop 1`] = `
"<svg id=\\"myId\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\" width=\\"135px\\" height=\\"41px\\" viewBox=\\"74 55 135 41\\">
  <style> #myId { pointer-events:none; } #myId .event  { pointer-events:all; } line { stroke-linecap:round; } polygon { stroke-linejoin:round; } </style>
  <text x=\\"154.94630484541327\\" y=\\"73.66666666666667\\" font-family=\\" Helvetica\\" font-size=\\"14\\" fill=\\"rgb(255,13,13)\\">O</text>
  <text x=\\"92.59247577293368\\" y=\\"85.66666666666667\\" font-family=\\" Helvetica\\" font-size=\\"14\\" fill=\\"rgb(255,13,13)\\">O</text>
  <text x=\\"82.48447577293368\\" y=\\"85.66666666666667\\" font-family=\\" Helvetica\\" font-size=\\"14\\" fill=\\"rgb(255,13,13)\\">H</text>
  <line x1=\\"201.96152422706632\\" y1=\\"69\\" x2=\\"181.1769145362398\\" y2=\\"81\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line x1=\\"181.1769145362398\\" y1=\\"81\\" x2=\\"166.83830484541326\\" y2=\\"72.72159983519633\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line x1=\\"153.94630484541327\\" y1=\\"72.72159983519633\\" x2=\\"139.60769515458674\\" y2=\\"81\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line x1=\\"139.60769515458674\\" y1=\\"81\\" x2=\\"118.82308546376021\\" y2=\\"69\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line x1=\\"118.82308546376021\\" y1=\\"69\\" x2=\\"104.48447577293368\\" y2=\\"77.27840016480367\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line id=\\"myId:Bond:0\\" class=\\"event\\" x1=\\"201.96152422706632\\" y1=\\"69\\" x2=\\"181.1769145362398\\" y2=\\"81\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <line id=\\"myId:Bond:1\\" class=\\"event\\" x1=\\"181.1769145362398\\" y1=\\"81\\" x2=\\"160.39230484541326\\" y2=\\"69\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <line id=\\"myId:Bond:2\\" class=\\"event\\" x1=\\"160.39230484541326\\" y1=\\"69\\" x2=\\"139.60769515458674\\" y2=\\"81\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <line id=\\"myId:Bond:3\\" class=\\"event\\" x1=\\"139.60769515458674\\" y1=\\"81\\" x2=\\"118.82308546376021\\" y2=\\"69\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <line id=\\"myId:Bond:4\\" class=\\"event\\" x1=\\"118.82308546376021\\" y1=\\"69\\" x2=\\"98.03847577293368\\" y2=\\"81\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:0\\" class=\\"event\\" cx=\\"201.96152422706632\\" cy=\\"69\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:1\\" class=\\"event\\" cx=\\"181.1769145362398\\" cy=\\"81\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:2\\" class=\\"event\\" cx=\\"160.39230484541326\\" cy=\\"69\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:3\\" class=\\"event\\" cx=\\"139.60769515458674\\" cy=\\"81\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:4\\" class=\\"event\\" cx=\\"118.82308546376021\\" cy=\\"69\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:5\\" class=\\"event\\" cx=\\"98.03847577293368\\" cy=\\"81\\" r=\\"8\\" fill-opacity=\\"0\\" />
</svg>"
`;

exports[`Molecule toSVG with autoCrop 2`] = `
"<svg id=\\"myId\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\" width=\\"175px\\" height=\\"81px\\" viewBox=\\"54 35 175 81\\">
  <style> #myId { pointer-events:none; } #myId .event  { pointer-events:all; } line { stroke-linecap:round; } polygon { stroke-linejoin:round; } </style>
  <text x=\\"154.94630484541327\\" y=\\"73.66666666666667\\" font-family=\\" Helvetica\\" font-size=\\"14\\" fill=\\"rgb(255,13,13)\\">O</text>
  <text x=\\"92.59247577293368\\" y=\\"85.66666666666667\\" font-family=\\" Helvetica\\" font-size=\\"14\\" fill=\\"rgb(255,13,13)\\">O</text>
  <text x=\\"82.48447577293368\\" y=\\"85.66666666666667\\" font-family=\\" Helvetica\\" font-size=\\"14\\" fill=\\"rgb(255,13,13)\\">H</text>
  <line x1=\\"201.96152422706632\\" y1=\\"69\\" x2=\\"181.1769145362398\\" y2=\\"81\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line x1=\\"181.1769145362398\\" y1=\\"81\\" x2=\\"166.83830484541326\\" y2=\\"72.72159983519633\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line x1=\\"153.94630484541327\\" y1=\\"72.72159983519633\\" x2=\\"139.60769515458674\\" y2=\\"81\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line x1=\\"139.60769515458674\\" y1=\\"81\\" x2=\\"118.82308546376021\\" y2=\\"69\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line x1=\\"118.82308546376021\\" y1=\\"69\\" x2=\\"104.48447577293368\\" y2=\\"77.27840016480367\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line id=\\"myId:Bond:0\\" class=\\"event\\" x1=\\"201.96152422706632\\" y1=\\"69\\" x2=\\"181.1769145362398\\" y2=\\"81\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <line id=\\"myId:Bond:1\\" class=\\"event\\" x1=\\"181.1769145362398\\" y1=\\"81\\" x2=\\"160.39230484541326\\" y2=\\"69\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <line id=\\"myId:Bond:2\\" class=\\"event\\" x1=\\"160.39230484541326\\" y1=\\"69\\" x2=\\"139.60769515458674\\" y2=\\"81\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <line id=\\"myId:Bond:3\\" class=\\"event\\" x1=\\"139.60769515458674\\" y1=\\"81\\" x2=\\"118.82308546376021\\" y2=\\"69\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <line id=\\"myId:Bond:4\\" class=\\"event\\" x1=\\"118.82308546376021\\" y1=\\"69\\" x2=\\"98.03847577293368\\" y2=\\"81\\" stroke-width=\\"8\\" stroke-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:0\\" class=\\"event\\" cx=\\"201.96152422706632\\" cy=\\"69\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:1\\" class=\\"event\\" cx=\\"181.1769145362398\\" cy=\\"81\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:2\\" class=\\"event\\" cx=\\"160.39230484541326\\" cy=\\"69\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:3\\" class=\\"event\\" cx=\\"139.60769515458674\\" cy=\\"81\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:4\\" class=\\"event\\" cx=\\"118.82308546376021\\" cy=\\"69\\" r=\\"8\\" fill-opacity=\\"0\\" />
  <circle id=\\"myId:Atom:5\\" class=\\"event\\" cx=\\"98.03847577293368\\" cy=\\"81\\" r=\\"8\\" fill-opacity=\\"0\\" />
</svg>"
`;
